Bump onnx.Constant to opset 21, adding int4 and uint4 support

jorickert opened the pull request:
Requires https://github.com/Xilinx/llvm-project/pull/568

jorickert committed:
Bump onnx.Constant to opset 21, adding int4 and uint4 support

Signed-off-by: Rickert, Jonas <Jonas.Rickert@amd.com>

jorickert commented on test/mlir/onnx/parse/int4_nonraw_data.json:
I would have prefered to use .onnxtxt instead of the json representation here, but due to https://github.com/onnx/onnx/issues/6152 this is not possible.

please pay extra attention to the tests, as I wrote both the tests and the parsing myself, it could be that there is a common error in both the test inputs and the parsing logic

jorickert committed:
Fix tests for int4. Requires llvm bump

Signed-off-by: Rickert, Jonas <Jonas.Rickert@amd.com>

jorickert commented:
requires https://github.com/Xilinx/llvm-project/pull/568 , maybe the llvm commit in this repo may need to be updated.
If you have another approach, that does not require changes to LLVM, feel free to comment

roberteg16 committed:
chore: Rename inAnyInt4Type to isAnyInt4Type

roberteg16 committed:
chore: Fix wrong comment mixing LSB and MSB

roberteg16 committed:
chore: Add whitespaces

roberteg16 committed:
chore: Add missing test name

roberteg16 committed:
fix: Fix test to use integers instead of floats

roberteg16 committed:
chore: Update llvm-project commit hash

roberteg16 commented on test/mlir/onnx/parse/int4_raw_data.json:
```diff
@@ -0,0 +1,92 @@
+              ],
+              "dataType": 22,
+              "name": "tensor_int4",
+              "rawData": "EC8O"
```
I could not understand how rawData works, but to other test is correct

jorickert replied:
https://github.com/postrational/onnx/blob/9fb11e344a7721b2eed1f3e26bf9312f168a79a0/onnx/onnx.proto3#L625

External data (for which my PR was missing support) is very similar to rawData

jorickert committed:
Add/Fix support for external (u)int4 data.

Add a test for external int4 data.

Adjust test_splat to work with APInt and existing llvm APIs

Signed-off-by: Rickert, Jonas <Jonas.Rickert@amd.com>

jorickert committed:
Add (u)int4 to NonContiguousIterableTypesT

Signed-off-by: Rickert, Jonas <Jonas.Rickert@amd.com>
